File: backend/services/pathfinding-api/src/main/java/com/navigation/pathfindingapi/domain/PathfindingOptimizations.java
package com.navigation.pathfindingapi.domain;

import com.navigation.pathfinder.weight.DistanceEdgeWeightCalculator;
import com.navigation.pathfinder.weight.DurationEdgeWeightCalculator;
import com.navigation.pathfinder.weight.EdgeWeightCalculator;
import com.navigation.pathfinder.weight.VertexCountEdgeWeightCalculator;

import java.util.function.Supplier;

public enum PathfindingOptimizations {
  DISTANCE(DistanceEdgeWeightCalculator::new),
  NUMBER_OF_NODES(VertexCountEdgeWeightCalculator::new),
  TIME(DurationEdgeWeightCalculator::new);

  public final Supplier<? extends EdgeWeightCalculator> supplier;

  <T extends EdgeWeightCalculator> PathfindingOptimizations(Supplier<T> supplier) {
    this.supplier = supplier;
  }
}

File: client/src/constants/pathfinding-optimizations.ts
export enum OptimizationTypes {
  DISTANCE = 'DISTANCE',
  TIME = 'TIME',
  NUMBER_OF_NODES = 'NUMBER_OF_NODES',
  NONE = 'NONE',
}

export const OPTIMIZATION_TYPES_TRANSLATIONS = {
  [OptimizationTypes.TIME]: 'Time',
  [OptimizationTypes.DISTANCE]: 'Distance',
  [OptimizationTypes.NUMBER_OF_NODES]: 'Number of nodes',
  [OptimizationTypes.NONE]: 'None',
};
